kong/db/migrations/core/015_280_to_300.lua

```diff
@@ -0,0 +1,19 @@
+return {
+    postgres = {
+      up = [[
```

Comment:
currently the database stores all the known state for each DP: `last_seen`, `ip`, `config_hash`, `version`, `sync_status`... adding the negotiated state seems natural.  Also, given the limitation from lua-resty-websocket, which prevents from reusing the connection between negotiation and websocket, it's very likely that they would be served by different workers.  The safest way to carry this state between these two requests is to store in the database.  We could also use a shared dict, but I doubt it would be significantly better.

Reply:
>Also, given the limitation from lua-resty-websocket, which prevents from reusing the connection between negotiation and websocket, it's very likely that they would be served by different workers. The safest way to carry this state between these two requests is to store in the database. We could also use a shared dict, but I doubt it would be significantly better.

This is exactly what I'm concerned about. If the persisted state is used later on outside the lifecycle of a TCP connection, it implies coordination between control-plane nodes since the design allows for load-balancing of multiple connections amongst a set of nodes.

Reply:
isn't that a good reason precisely to store in the database?

Reply:
I think we should avoid any coordination vai the database at this level because:
- added latency to setup the wRPC state, very relevant since kong does polling and not a notify based event propagation
- wRPC's scope is limited to TCP connection and nothing outside it, which obviates away cross-node synchronization and consistency issues. This is very much intentional.

Reply:
there are two levels of cache between Lua code and the actual database, it's not likely to add significant latency unless it has time to get cold.
should we fork lua-resty-websocket to be able to use the same TLS connection?  i think yes; but not before Kong 3.0

Reply:
> there are two levels of cache between Lua code and the actual database, it's not likely to add significant latency unless it has time to get cold.

If the request lands in the same node, sure. If the request lands on another node, that's a whole another thing since the cache is lazily built.

>  should we fork lua-resty-websocket to be able to use the same TLS connection? i think yes; but not before Kong 3.0

Absolutely. Please track this work item so that we do not forget it.

Reply:
_NOTE/FYI:_ we already have [a fork of lua-resty-websocket](https://github.com/Kong/lua-resty-websocket) for additional features, and I am working to get it up on luarocks soon for easy inclusion into Kong :)